credentials/src/main/kotlin/web5/sdk/credentials/VerifiablePresentation.kt

```diff
@@ -0,0 +1,271 @@
+     *
+     * @param vpJwt The Verifiable Presentation in JWT format as a [String].
+     * @throws SignatureException if the verification fails at any step, providing a message with failure details.
+     * @throws IllegalArgumentException if critical JWT header elements are absent.
+     *
+     * ### Example:
+     * ```
+     * try {
+     *     VerifiablePresentation.verify(signedVpJwt)
+     *     println("VP Verification successful!")
+     * } catch (e: SignatureException) {
+     *     println("VP Verification failed: ${e.message}")
+     * }
+     * ```
+     */
+    public fun verify(vpJwt: String) {
+      val jwt = JWTParser.parse(vpJwt) as SignedJWT // validates JWT
+
+      require(jwt.header.algorithm != null && jwt.header.keyID != null) {
+        "Signature verification failed: Expected JWS header to contain alg and kid"
+      }
+
+      val verificationMethodId = jwt.header.keyID
+      val parsedDidUrl = DIDURL.fromString(verificationMethodId) // validates vm id which is a DID URL
+
+      val didResolutionResult = DidResolvers.resolve(parsedDidUrl.did.didString)
+      if (didResolutionResult.didResolutionMetadata?.error != null) {
+        throw SignatureException(
+          "Signature verification failed: " +
+            "Failed to resolve DID ${parsedDidUrl.did.didString}. " +
+            "Error: ${didResolutionResult.didResolutionMetadata?.error}"
+        )
+      }
+
+      // create a set of possible id matches. the DID spec allows for an id to be the entire `did#fragment`
+      // or just `#fragment`. See: https://www.w3.org/TR/did-core/#relative-did-urls.
+      // using a set for fast string comparison. DIDs can be lonnng.
+      val verificationMethodIds = setOf(parsedDidUrl.didUrlString, "#${parsedDidUrl.fragment}")
+      val assertionMethods = didResolutionResult.didDocument.assertionMethodVerificationMethodsDereferenced
+      val assertionMethod = assertionMethods?.firstOrNull {
+        val id = it.id.toString()
+        verificationMethodIds.contains(id)
+      }
+
+      if (assertionMethod == null) {
+        throw SignatureException(
+          "Signature verification failed: Expected kid in JWS header to dereference " +
+            "a DID Document Verification Method with an Assertion verification relationship"
+        )
+      }
+
+      require(assertionMethod.isType("JsonWebKey2020") && assertionMethod.publicKeyJwk != null) {
+        throw SignatureException(
+          "Signature verification failed: Expected kid in JWS header to dereference " +
+            "a DID Document Verification Method of type JsonWebKey2020 with a publicKeyJwk"
+        )
+      }
+
+      val publicKeyMap = assertionMethod.publicKeyJwk
+      val publicKeyJwk = JWK.parse(publicKeyMap)
+
+      val toVerifyBytes = jwt.signingInput
+      val signatureBytes = jwt.signature.decode()
+
+      Crypto.verify(publicKeyJwk, toVerifyBytes, signatureBytes, jwt.header.algorithm)
```

Comment:
Would it make sense to refactor this so logic is shared with the VC.verify function?

Reply:
Added